velox/expression/SimpleFunctionRegistry.h

```diff
@@ -55,9 +55,43 @@ struct FunctionEntry {
   const FunctionFactory factory_;
 };
 
+struct SignatureHash {
+  size_t operator()(const FunctionSignature& signature) const {
+    return std::hash<std::vector<TypeSignature>>()(signature.argumentTypes());
+  }
+};
+
+/// Only consider function signature argument types baseName.
+/// E.g. registered map<int, long> as function input argument before, now we
+/// try to register map<long, long>, we will get it from SignatureMap, but its
+/// physical type is not equal, so not overwrite map<int, long>, both map<int,
+/// long> and map<long, long> exist in FunctionEntry.
+/// Function with Variadic<int64_t> or int64_t should not exist at the same
+/// time, we cannot distinguish which to call for single input.
+struct SignatureCmp {
+  bool operator()(const FunctionSignature& l, const FunctionSignature& r)
+      const {
+    if (l.argumentTypes().size() != r.argumentTypes().size()) {
+      return false;
+    }
+    for (auto i = 0; i < l.argumentTypes().size(); ++i) {
+      // For Map(K, V), we only consider the argument type map, not consider its
+      // element type, so as Decimal, List, Struct.let the metadata
+      // physicalSignatureEquals which will consider the child data type to
+      // decide if overwrite.
+      if (l.argumentTypes()[i].baseName() != r.argumentTypes()[i].baseName()) {
```

Comment:
I don't think this works. We cannot store 2 entries in a map if their keys compare as equal. Hence, we won't be able to store foo(array(bigint)) and foo(array(boolean)) at the same.

Reply:
IN the signature, we use `array(T)` to represents `array(bigint)` and `array(boolean)`, so the baseName is equal.
Suppose we have registered `array(bigint)`, now we try to register `array<boolean>`, the `functions` size is 1, then the `metadata->physicalSignatureEquals(otherMetadata)` condition is false, so we can insert this FunctionEntry `array(boolean)`.
https://github.com/facebookincubator/velox/blob/main/velox/expression/SimpleFunctionRegistry.cpp#L44
https://github.com/facebookincubator/velox/blob/main/velox/expression/SimpleFunctionRegistry.cpp#L49

Reply:
Yes, so I have changed to both the SignatureMap and SignatureCmp only consider `baseName`. It is equal.